File: l0415_add_strings/add_strings.go
package leetcode0415

import "strconv"

// Solution 1: Simulate
// Time: O(max(m, n))
// Space: O(1)
func addStrings(num1 string, num2 string) string {
	res := ""
	carry := 0

	for i, j := len(num1)-1, len(num2)-1; i >= 0 || j >= 0 || carry != 0; i, j = i-1, j-1 {
		var x, y int
		if i >= 0 {
			x = int(num1[i] - '0')
		}
		if j >= 0 {
			y = int(num2[j] - '0')
		}
		sum := x + y + carry
		res = strconv.Itoa(sum%10) + res
		carry = sum / 10
	}

	return res
}

File: l0415_add_strings/add_strings.c
#include <string.h>

char* addStrings(char* num1, char* num2) {
    int i = strlen(num1) - 1;
    int j = strlen(num2) - 1;
    int carry = 0;
    // 1 for array size, 1 for carry, 1 for '\0'
    char *res = (char*)malloc(sizeof(char) * (max(i, j) + 3));
    int len = 0;

    while (i >= 0 || j >= 0 || carry != 0) {
        int x = i >= 0 ? num1[i] - '0' : 0;
        int y = j >= 0 ? num2[j] - '0' : 0;
        int sum = x + y + carry;
        res[len++] = '0' + sum % 10;
        carry = sum / 10;
        i--;
        j--;
    }
    for (int i = 0; 2 * i < len; i++) {
        int t = res[i];
        res[i] = res[len - i - 1], res[len - i - 1] = t;
    }
    res[len++] = 0;
    return res;
}

int max(int i, int j) {
    if (i > j) {
        return i;
    }
    return j;
}